omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+	if err != nil {
+		panic(err)
+	}
+
+	n := len(sb.Roster.List)
+	cosiProto := proto.(*cosiprotocol.FtCosi)
+	cosiProto.Msg = req.Hash()
+	cosiProto.Data = payload
+	cosiProto.CreateProtocol = s.CreateProtocol
+	cosiProto.Timeout = interval / 2
+	cosiProto.Threshold = n - n/3
+	cosiProto.NSubtrees = int(math.Pow(float64(n), 1.0/3.0))
+	if err := cosiProto.Start(); err != nil {
+		return nil, err
+	}
+	// The protocol should always send FinalSignature because it has a
+	// timeout, so we don't need a select.
+	return <-cosiProto.FinalSignature, nil
+}
+
+// verifyViewChange is registered in the view-change ftcosi.
+func (s *Service) verifyViewChange(msg []byte, data []byte) bool {
+	// Parse message and check hash.
+	var req viewchange.NewViewReq
+	if err := protobuf.DecodeWithConstructors(data, &req, network.DefaultConstructors(cothority.Suite)); err != nil {
+		log.Error(s.ServerIdentity(), err)
+		return false
+	}
+	if !bytes.Equal(msg, req.Hash()) {
+		log.Error(s.ServerIdentity(), "digest doesn't verify")
+		return false
+	}
+	// Check that we know about the view and the new roster in the request
+	// matches the view-change proofs.
+	sb := s.db().GetByID(req.GetView().ID)
+	if sb == nil {
+		log.Error(s.ServerIdentity(), "view does not exist")
+		return false
+	}
+	newRosterID := rotateRoster(sb.Roster, req.GetView().LeaderIndex).ID
+	if !newRosterID.Equal(req.Roster.ID) {
+		log.Error(s.ServerIdentity(), "invalid roster in request")
+		return false
+	}
+	// Check the signers are unique, they are in the roster and the
+	// signatures are correct.
+	uniqueSigners, uniqueViews := func() (int, int) {
+		signers := make(map[[16]byte]bool)
+		views := make(map[string]bool)
+		for _, p := range req.Proof {
+			signers[p.SignerID] = true
+			views[string(p.View.Hash())] = true
+		}
+		return len(signers), len(views)
+	}()
+	f := len(sb.Roster.List) / 3
+	if uniqueSigners < 2*f+1 {
+		log.Error(s.ServerIdentity(), "not enough proofs")
+		return false
+	}
+	if uniqueViews != 1 {
+		log.Error(s.ServerIdentity(), "conflicting views")
+		return false
+	}
+	for _, p := range req.Proof {
+		_, sid := sb.Roster.Search(p.SignerID)
+		if sid == nil {
+			log.Error(s.ServerIdentity(), "the signer is not in the roster")
+			return false
+		}
+		// Check that the signature is correct.
+		if err := schnorr.Verify(cothority.Suite, sid.Public, p.Hash(), p.Signature); err != nil {
+			log.Error(s.ServerIdentity(), err)
+			return false
+		}
+	}
+	log.Lvl2(s.ServerIdentity(), "view-change verification OK")
+	return true
+}
+
+// createViewChangeBlock creates a new block to record the successful
+// view-change operation.
+func (s *Service) createViewChangeBlock(req viewchange.NewViewReq, multisig []byte) error {
+	defer log.Lvl2(s.ServerIdentity(), "created view-change block")
+	sb, err := s.db().GetLatestByID(req.GetGen())
+	if err != nil {
+		return err
+	}
+	if len(sb.Roster.List) < 2 {
+		return errors.New("roster size is too small")
```

Comment:
Is this really the first place that view-change for 2 node cothorities is refused? A little too late and hard to understand isn't it? Someone trying to set up a too-small OmniLedger should get a nice clear message earlier. (File a new issue on this if you want, it's not exactly specific to this commit.)

Reply:
Yeah, we should make a new issue for that. View-change shouldn't work for roster smaller than 4 because it won't satisfy the n > 2f+1 requirement.